Utilisation locale simplifiée pour suivi non-technique

## Problème / motivation
Permettre aux membres non-techniques de l'équipe (@johanricher, @DaFrenchFrog) de : 

* Voir l'état actuel de l'application avec un minimum de connaissances / compétences sur la stack technique
* Pour éventuellement être autonome pour des modifications de code simples (texte, ...)

## Solution envisagée

Un setup `docker-compose` pour front + api + db (+ tout autre élément), à lancer en une seule commande

## Alternatives considérées

- Ne rien changer, les membres non-techniques suivent la doc actuelle : à évaluer, risque de devoir gérer des pb techniques intempestifs (configurer un postgres local ? -> pour info il y a déjà un `docker-compose` qui en démarre une)
- Environnement de démo (#67) uniquement : c'est peut-être pertinent mais orthogonal, par ex ça ne permettrait pas une prise en main autonome pour des modifs simples

@florimondmanca voici les erreurs successives que j'ai.

Make install
```
(venv) RomainG@RomainG catalogage-donnees % make install
python3 -m venv venv
venv/bin/pip install -U pip wheel setuptools
Requirement already satisfied: pip in ./venv/lib/python3.9/site-packages (22.0.3)
Requirement already satisfied: wheel in ./venv/lib/python3.9/site-packages (0.37.1)
Requirement already satisfied: setuptools in ./venv/lib/python3.9/site-packages (60.9.0)
Collecting setuptools
  Downloading setuptools-60.9.2-py3-none-any.whl (1.1 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 3.3 MB/s eta 0:00:00
Installing collected packages: setuptools
  Attempting uninstall: setuptools
    Found existing installation: setuptools 60.9.0
    Uninstalling setuptools-60.9.0:
      Successfully uninstalled setuptools-60.9.0
Successfully installed setuptools-60.9.2
venv/bin/pip install -r requirements.txt
Requirement already satisfied: asyncpg==0.25.0 in ./venv/lib/python3.9/site-packages (from -r requirements.txt (line 2)) (0.25.0)
Requirement already satisfied: alembic==1.7.5 in ./venv/lib/python3.9/site-packages (from -r requirements.txt (line 3)) (1.7.5)
Requirement already satisfied: fastapi==0.70.0 in ./venv/lib/python3.9/site-packages (from -r requirements.txt (line 4)) (0.70.0)
Requirement already satisfied: gunicorn==20.1.0 in ./venv/lib/python3.9/site-packages (from -r requirements.txt (line 5)) (20.1.0)
Requirement already satisfied: punq==0.5 in ./venv/lib/python3.9/site-packages (from -r requirements.txt (line 6)) (0.5)
Requirement already satisfied: pydantic[email]==1.8.2 in ./venv/lib/python3.9/site-packages (from -r requirements.txt (line 7)) (1.8.2)
Requirement already satisfied: uvicorn[standard]==0.16.0 in ./venv/lib/python3.9/site-packages (from -r requirements.txt (line 8)) (0.16.0)
Requirement already satisfied: sqlalchemy[asyncio,mypy]==1.4.27 in ./venv/lib/python3.9/site-packages (from -r requirements.txt (line 9)) (1.4.27)
Requirement already satisfied: asgi-lifespan==1.0.1 in ./venv/lib/python3.9/site-packages (from -r requirements.txt (line 12)) (1.0.1)
Requirement already satisfied: black==21.12b0 in ./venv/lib/python3.9/site-packages (from -r requirements.txt (line 13)) (21.12b0)
Requirement already satisfied: flake8==4.0.1 in ./venv/lib/python3.9/site-packages (from -r requirements.txt (line 14)) (4.0.1)
Requirement already satisfied: httpx==0.21.1 in ./venv/lib/python3.9/site-packages (from -r requirements.txt (line 15)) (0.21.1)
Requirement already satisfied: isort==5.10.1 in ./venv/lib/python3.9/site-packages (from -r requirements.txt (line 16)) (5.10.1)
Requirement already satisfied: mypy==0.910 in ./venv/lib/python3.9/site-packages (from -r requirements.txt (line 17)) (0.910)
Requirement already satisfied: psycopg2-binary==2.9.2 in ./venv/lib/python3.9/site-packages (from -r requirements.txt (line 18)) (2.9.2)
Requirement already satisfied: pytest==6.2.5 in ./venv/lib/python3.9/site-packages (from -r requirements.txt (line 19)) (6.2.5)
Requirement already satisfied: pytest-asyncio==0.16.0 in ./venv/lib/python3.9/site-packages (from -r requirements.txt (line 20)) (0.16.0)
Requirement already satisfied: pytest-cov==3.0.0 in ./venv/lib/python3.9/site-packages (from -r requirements.txt (line 21)) (3.0.0)
Requirement already satisfied: sqlalchemy-utils==0.37.9 in ./venv/lib/python3.9/site-packages (from -r requirements.txt (line 22)) (0.37.9)
Requirement already satisfied: Mako in ./venv/lib/python3.9/site-packages (from alembic==1.7.5->-r requirements.txt (line 3)) (1.1.6)
Requirement already satisfied: starlette==0.16.0 in ./venv/lib/python3.9/site-packages (from fastapi==0.70.0->-r requirements.txt (line 4)) (0.16.0)
Requirement already satisfied: setuptools>=3.0 in ./venv/lib/python3.9/site-packages (from gunicorn==20.1.0->-r requirements.txt (line 5)) (60.9.2)
Requirement already satisfied: typing-extensions>=3.7.4.3 in ./venv/lib/python3.9/site-packages (from pydantic[email]==1.8.2->-r requirements.txt (line 7)) (4.1.1)
Requirement already satisfied: email-validator>=1.0.3 in ./venv/lib/python3.9/site-packages (from pydantic[email]==1.8.2->-r requirements.txt (line 7)) (1.1.3)
Requirement already satisfied: h11>=0.8 in ./venv/lib/python3.9/site-packages (from uvicorn[standard]==0.16.0->-r requirements.txt (line 8)) (0.12.0)
Requirement already satisfied: asgiref>=3.4.0 in ./venv/lib/python3.9/site-packages (from uvicorn[standard]==0.16.0->-r requirements.txt (line 8)) (3.5.0)
Requirement already satisfied: click>=7.0 in ./venv/lib/python3.9/site-packages (from uvicorn[standard]==0.16.0->-r requirements.txt (line 8)) (8.0.3)
Requirement already satisfied: PyYAML>=5.1 in ./venv/lib/python3.9/site-packages (from uvicorn[standard]==0.16.0->-r requirements.txt (line 8)) (6.0)
Requirement already satisfied: websockets>=10.0 in ./venv/lib/python3.9/site-packages (from uvicorn[standard]==0.16.0->-r requirements.txt (line 8)) (10.1)
Requirement already satisfied: httptools<0.4.0,>=0.2.0 in ./venv/lib/python3.9/site-packages (from uvicorn[standard]==0.16.0->-r requirements.txt (line 8)) (0.3.0)
Requirement already satisfied: watchgod>=0.6 in ./venv/lib/python3.9/site-packages (from uvicorn[standard]==0.16.0->-r requirements.txt (line 8)) (0.7)
Requirement already satisfied: python-dotenv>=0.13 in ./venv/lib/python3.9/site-packages (from uvicorn[standard]==0.16.0->-r requirements.txt (line 8)) (0.19.2)
Requirement already satisfied: uvloop!=0.15.0,!=0.15.1,>=0.14.0 in ./venv/lib/python3.9/site-packages (from uvicorn[standard]==0.16.0->-r requirements.txt (line 8)) (0.16.0)
Requirement already satisfied: greenlet!=0.4.17 in ./venv/lib/python3.9/site-packages (from sqlalchemy[asyncio,mypy]==1.4.27->-r requirements.txt (line 9)) (1.1.2)
Requirement already satisfied: sqlalchemy2-stubs in ./venv/lib/python3.9/site-packages (from sqlalchemy[asyncio,mypy]==1.4.27->-r requirements.txt (line 9)) (0.0.2a19)
Requirement already satisfied: sniffio in ./venv/lib/python3.9/site-packages (from asgi-lifespan==1.0.1->-r requirements.txt (line 12)) (1.2.0)
Requirement already satisfied: mypy-extensions>=0.4.3 in ./venv/lib/python3.9/site-packages (from black==21.12b0->-r requirements.txt (line 13)) (0.4.3)
Requirement already satisfied: platformdirs>=2 in ./venv/lib/python3.9/site-packages (from black==21.12b0->-r requirements.txt (line 13)) (2.5.0)
Requirement already satisfied: pathspec<1,>=0.9.0 in ./venv/lib/python3.9/site-packages (from black==21.12b0->-r requirements.txt (line 13)) (0.9.0)
Requirement already satisfied: tomli<2.0.0,>=0.2.6 in ./venv/lib/python3.9/site-packages (from black==21.12b0->-r requirements.txt (line 13)) (1.2.3)
Requirement already satisfied: pyflakes<2.5.0,>=2.4.0 in ./venv/lib/python3.9/site-packages (from flake8==4.0.1->-r requirements.txt (line 14)) (2.4.0)
Requirement already satisfied: mccabe<0.7.0,>=0.6.0 in ./venv/lib/python3.9/site-packages (from flake8==4.0.1->-r requirements.txt (line 14)) (0.6.1)
Requirement already satisfied: pycodestyle<2.9.0,>=2.8.0 in ./venv/lib/python3.9/site-packages (from flake8==4.0.1->-r requirements.txt (line 14)) (2.8.0)
Requirement already satisfied: certifi in ./venv/lib/python3.9/site-packages (from httpx==0.21.1->-r requirements.txt (line 15)) (2021.10.8)
Requirement already satisfied: charset-normalizer in ./venv/lib/python3.9/site-packages (from httpx==0.21.1->-r requirements.txt (line 15)) (2.0.12)
Requirement already satisfied: rfc3986[idna2008]<2,>=1.3 in ./venv/lib/python3.9/site-packages (from httpx==0.21.1->-r requirements.txt (line 15)) (1.5.0)
Requirement already satisfied: httpcore<0.15.0,>=0.14.0 in ./venv/lib/python3.9/site-packages (from httpx==0.21.1->-r requirements.txt (line 15)) (0.14.7)
Requirement already satisfied: toml in ./venv/lib/python3.9/site-packages (from mypy==0.910->-r requirements.txt (line 17)) (0.10.2)
Requirement already satisfied: pluggy<2.0,>=0.12 in ./venv/lib/python3.9/site-packages (from pytest==6.2.5->-r requirements.txt (line 19)) (1.0.0)
Requirement already satisfied: packaging in ./venv/lib/python3.9/site-packages (from pytest==6.2.5->-r requirements.txt (line 19)) (21.3)
Requirement already satisfied: py>=1.8.2 in ./venv/lib/python3.9/site-packages (from pytest==6.2.5->-r requirements.txt (line 19)) (1.11.0)
Requirement already satisfied: iniconfig in ./venv/lib/python3.9/site-packages (from pytest==6.2.5->-r requirements.txt (line 19)) (1.1.1)
Requirement already satisfied: attrs>=19.2.0 in ./venv/lib/python3.9/site-packages (from pytest==6.2.5->-r requirements.txt (line 19)) (21.4.0)
Requirement already satisfied: coverage[toml]>=5.2.1 in ./venv/lib/python3.9/site-packages (from pytest-cov==3.0.0->-r requirements.txt (line 21)) (6.3.1)
Requirement already satisfied: six in ./venv/lib/python3.9/site-packages (from sqlalchemy-utils==0.37.9->-r requirements.txt (line 22)) (1.16.0)
Requirement already satisfied: anyio<4,>=3.0.0 in ./venv/lib/python3.9/site-packages (from starlette==0.16.0->fastapi==0.70.0->-r requirements.txt (line 4)) (3.5.0)
Requirement already satisfied: dnspython>=1.15.0 in ./venv/lib/python3.9/site-packages (from email-validator>=1.0.3->pydantic[email]==1.8.2->-r requirements.txt (line 7)) (2.2.0)
Requirement already satisfied: idna>=2.0.0 in ./venv/lib/python3.9/site-packages (from email-validator>=1.0.3->pydantic[email]==1.8.2->-r requirements.txt (line 7)) (3.3)
Requirement already satisfied: MarkupSafe>=0.9.2 in ./venv/lib/python3.9/site-packages (from Mako->alembic==1.7.5->-r requirements.txt (line 3)) (2.0.1)
Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in ./venv/lib/python3.9/site-packages (from packaging->pytest==6.2.5->-r requirements.txt (line 19)) (3.0.7)
cd client && npm ci && npx playwright install firefox
npm ERR! esbuild-android-arm64 not accessible from esbuild

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/user/.npm/_logs/2022-02-17T09_23_26_586Z-debug.log
make: *** [install-client] Error 1
```

Make migrate

```
(venv) RomainG@RomainG catalogage-donnees % make migrate
venv/bin/alembic upgrade head
Traceback (most recent call last):
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/bin/alembic", line 8, in <module>
    sys.exit(main())
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/alembic/config.py", line 588, in main
    CommandLine(prog=prog).main(argv=argv)
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/alembic/config.py", line 582, in main
    self.run_cmd(cfg, options)
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/alembic/config.py", line 559, in run_cmd
    fn(
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/alembic/command.py", line 320, in upgrade
    script.run_env()
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/alembic/script/base.py", line 563, in run_env
    util.load_python_file(self.dir, "env.py")
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/alembic/util/pyfiles.py", line 92, in load_python_file
    module = load_module_py(module_id, path)
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/alembic/util/pyfiles.py", line 108, in load_module_py
    spec.loader.exec_module(module)  # type: ignore
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "server/migrations/env.py", line 100, in <module>
    asyncio.run(run_migrations_online())
  File "/usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "server/migrations/env.py", line 93, in run_migrations_online
    async with connectable.connect() as connection:
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/ext/asyncio/base.py", line 60, in __aenter__
    return await self.start(is_ctxmanager=True)
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/ext/asyncio/engine.py", line 131, in start
    await (greenlet_spawn(self.sync_engine.connect))
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 134, in greenlet_spawn
    result = context.throw(*sys.exc_info())
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/future/engine.py", line 406, in connect
    return super(Engine, self).connect()
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 3197, in connect
    return self._connection_cls(self, close_with_result=close_with_result)
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 96, in __init__
    else engine.raw_connection()
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 3276, in raw_connection
    return self._wrap_pool_connect(self.pool.connect, _connection)
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 3243, in _wrap_pool_connect
    return fn()
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 310, in connect
    return _ConnectionFairy._checkout(self)
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 868, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 476, in checkout
    rec = pool._do_get()
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/pool/impl.py", line 256, in _do_get
    return self._create_connection()
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 256, in _create_connection
    return _ConnectionRecord(self)
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 371, in __init__
    self.__connect()
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 666, in __connect
    pool.logger.debug("Error on connect(): %s", e)
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 661, in __connect
    self.dbapi_connection = connection = pool._invoke_creator(self)
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/engine/create.py", line 590, in connect
    return dialect.connect(*cargs, **cparams)
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 584, in connect
    return self.dbapi.connect(*cargs, **cparams)
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 755, in connect
    await_only(self.asyncpg.connect(*arg, **kw)),
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 76, in await_only
    return current.driver.switch(awaitable)
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 129, in greenlet_spawn
    value = await result
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/asyncpg/connection.py", line 2085, in connect
    return await connect_utils._connect(
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/asyncpg/connect_utils.py", line 895, in _connect
    raise last_error
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/asyncpg/connect_utils.py", line 881, in _connect
    return await _connect_addr(
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/asyncpg/connect_utils.py", line 781, in _connect_addr
    return await __connect_addr(params, timeout, True, *args)
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/asyncpg/connect_utils.py", line 825, in __connect_addr
    tr, pr = await compat.wait_for(connector, timeout=timeout)
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/asyncpg/compat.py", line 66, in wait_for
    return await asyncio.wait_for(fut, timeout)
  File "/usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/tasks.py", line 479, in wait_for
    return fut.result()
  File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/asyncpg/connect_utils.py", line 699, in _create_ssl_connection
    do_ssl_upgrade = await pr.on_data
ConnectionError: unexpected connection_lost() call
make: *** [migrate] Error 1
```

Make serve

```
(venv) RomainG@RomainG catalogage-donnees % make serve
make -j 2 serve-server serve-client
./tools/colorize_prefix.sh [server] 34 "venv/bin/python -m server.main"
./tools/colorize_prefix.sh [client] 33 "cd client && npm run dev"
[client] 
[client] > client@0.0.1 dev /Users/user/Documents/_DEV/Svelte/catalogage-donnees/client
[client] > svelte-kit dev
[client] 
[client] sh: svelte-kit: command not found
[client] npm ERR! code ELIFECYCLE
[client] npm ERR! syscall spawn
[client] npm ERR! file sh
[client] npm ERR! errno ENOENT
[client] npm ERR! client@0.0.1 dev: `svelte-kit dev`
[client] npm ERR! spawn ENOENT
[client] npm ERR! 
[client] npm ERR! Failed at the client@0.0.1 dev script.
[client] npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
[client] npm WARN Local package.json exists, but node_modules missing, did you mean to install?
[client] 
[client] npm ERR! A complete log of this run can be found in:
[client] npm ERR!     /Users/user/.npm/_logs/2022-02-17T09_24_04_941Z-debug.log
[client] [server] INFO:     Will watch for changes in these directories: ['/Users/user/Documents/_DEV/Svelte/catalogage-donnees/server']
[server] INFO:     Uvicorn running on http://127.0.0.1:3579 (Press CTRL+C to quit)
[server] INFO:     Started reloader process [66536] using watchgod
[server] INFO:     Started server process [66542]
[server] INFO:     Waiting for application startup.
[server] ERROR:    Traceback (most recent call last):
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/starlette/routing.py", line 621, in lifespan
[server]     async with self.lifespan_context(app):
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/starlette/routing.py", line 518, in __aenter__
[server]     await self._router.startup()
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/starlette/routing.py", line 598, in startup
[server]     await handler()
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/server/infrastructure/database.py", line 17, in create_all
[server]     async with self._engine.begin() as conn:
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/ext/asyncio/base.py", line 60, in __aenter__
[server]     return await self.start(is_ctxmanager=True)
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/ext/asyncio/engine.py", line 586, in start
[server]     await self.conn.start(is_ctxmanager=is_ctxmanager)
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/ext/asyncio/engine.py", line 131, in start
[server]     await (greenlet_spawn(self.sync_engine.connect))
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 134, in greenlet_spawn
[server]     result = context.throw(*sys.exc_info())
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/future/engine.py", line 406, in connect
[server]     return super(Engine, self).connect()
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 3197, in connect
[server]     return self._connection_cls(self, close_with_result=close_with_result)
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 96, in __init__
[server]     else engine.raw_connection()
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 3276, in raw_connection
[server]     return self._wrap_pool_connect(self.pool.connect, _connection)
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 3243, in _wrap_pool_connect
[server]     return fn()
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 310, in connect
[server]     return _ConnectionFairy._checkout(self)
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 868, in _checkout
[server]     fairy = _ConnectionRecord.checkout(pool)
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 476, in checkout
[server]     rec = pool._do_get()
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/pool/impl.py", line 146, in _do_get
[server]     self._dec_overflow()
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
[server]     compat.raise_(
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
[server]     raise exception
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/pool/impl.py", line 143, in _do_get
[server]     return self._create_connection()
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 256, in _create_connection
[server]     return _ConnectionRecord(self)
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 371, in __init__
[server]     self.__connect()
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 666, in __connect
[server]     pool.logger.debug("Error on connect(): %s", e)
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
[server]     compat.raise_(
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
[server]     raise exception
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 661, in __connect
[server]     self.dbapi_connection = connection = pool._invoke_creator(self)
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/engine/create.py", line 590, in connect
[server]     return dialect.connect(*cargs, **cparams)
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 584, in connect
[server]     return self.dbapi.connect(*cargs, **cparams)
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 755, in connect
[server]     await_only(self.asyncpg.connect(*arg, **kw)),
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 76, in await_only
[server]     return current.driver.switch(awaitable)
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 129, in greenlet_spawn
[server]     value = await result
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/asyncpg/connection.py", line 2085, in connect
[server]     return await connect_utils._connect(
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/asyncpg/connect_utils.py", line 895, in _connect
[server]     raise last_error
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/asyncpg/connect_utils.py", line 881, in _connect
[server]     return await _connect_addr(
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/asyncpg/connect_utils.py", line 781, in _connect_addr
[server]     return await __connect_addr(params, timeout, True, *args)
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/asyncpg/connect_utils.py", line 825, in __connect_addr
[server]     tr, pr = await compat.wait_for(connector, timeout=timeout)
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/asyncpg/compat.py", line 66, in wait_for
[server]     return await asyncio.wait_for(fut, timeout)
[server]   File "/usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/tasks.py", line 479, in wait_for
[server]     return fut.result()
[server]   File "/Users/user/Documents/_DEV/Svelte/catalogage-donnees/venv/lib/python3.9/site-packages/asyncpg/connect_utils.py", line 699, in _create_ssl_connection
[server]     do_ssl_upgrade = await pr.on_data
[server] ConnectionError: unexpected connection_lost() call
[server] 
[server] ERROR:    Application startup failed. Exiting.
[server] INFO:     Stopping reloader process [66536]
```

--- a/.dockerignore
+++ b/.dockerignore
@@ -1,0 +1,8 @@
+.git/
+venv/
+.mypy_cache/
+.pytest_cache/
+**/node_modules/
+**/.vagrant/
+.vscode/
+.coverage

--- a/Makefile
+++ b/Makefile
@@ -35,10 +35,10 @@ serve-dist: #- Serve both the server and the built client in parallel
 serve-dist-client: #- Run the built client 
 	./tools/colorize_prefix.sh [client] 33 "cd client && npm start"
 
-up: #- Start Docker Compose setup
+compose-up: #- Start Docker Compose setup
 	docker-compose up --build -d
 
-down: #- Stop and teardown Docker Compose setup
+compose-down: #- Stop and teardown Docker Compose setup
 	docker-compose down
 
 migrate: #- Apply pending migrations

--- a/Makefile
+++ b/Makefile
@@ -37,6 +37,7 @@ serve-dist-client: #- Run the built client
 
 compose-up: #- Start Docker Compose setup
 	docker-compose up --build -d
+	docker-compose run migrate
 
 compose-down: #- Stop and teardown Docker Compose setup
 	docker-compose down

--- a/Makefile
+++ b/Makefile
@@ -35,6 +35,12 @@ serve-dist: #- Serve both the server and the built client in parallel
 serve-dist-client: #- Run the built client 
 	./tools/colorize_prefix.sh [client] 33 "cd client && npm start"
 
+up: #- Start Docker Compose setup
+	docker-compose up --build -d
+
+down: #- Stop and teardown Docker Compose setup
+	docker-compose down
+
 migrate: #- Apply pending migrations
 	${bin}alembic upgrade head
 

--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -21,7 +21,7 @@ services:
     ports:
       - "3000:3000"
     environment:
-      VITE_API_BROWSER_URL: "http://localhost:3000"
+      VITE_API_BROWSER_URL: "http://localhost:3579"
       VITE_API_SSR_URL: "http://server:3579"
     volumes:
       - "./client/src:/app/client/src"
@@ -31,13 +31,21 @@ services:
     build:
       context: .
       dockerfile: ./docker/server.Dockerfile
+    image: catalogage-server
     container_name: catalogage-server
     ports:
       - "3579:3579"
     environment:
       APP_HOST: "0.0.0.0"
-      APP_DATABASE_URL: "postgresql+asyncpg://user:pass@db/catalogage"
+      APP_DATABASE_URL: "postgresql+asyncpg://user:pass@db:5432/catalogage"
     restart: always
+
+  migrate:
+    image: catalogage-server
+    profiles: ["cli"]
+    command: ./tools/wait-for-it.sh db:5432 -- make migrate
+    environment:
+      APP_DATABASE_URL: "postgresql+asyncpg://user:pass@db:5432/catalogage"
 
 volumes:
   pgdata:

--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -1,10 +1,9 @@
-version: '3'
-
+version: "3"
 
 services:
-  postgresql:
+  db:
     image: postgres:12
-    container_name: catalogage-pg
+    container_name: catalogage-db
     ports:
       - "5432:5432"
     environment:
@@ -12,7 +11,33 @@ services:
       - POSTGRES_USER=user
       - POSTGRES_PASSWORD=pass
     volumes:
-      - pgdata:/var/lib/postgresql/data/ # persist data even if container shuts down
+      - pgdata:/var/lib/postgresql/data/
+
+  client:
+    build:
+      context: .
+      dockerfile: ./docker/client.Dockerfile
+    container_name: catalogage-client
+    ports:
+      - "3000:3000"
+    environment:
+      VITE_API_BROWSER_URL: "http://localhost:3000"
+      VITE_API_SSR_URL: "http://server:3579"
+    volumes:
+      - "./client/src:/app/client/src"
+    restart: always
+
+  server:
+    build:
+      context: .
+      dockerfile: ./docker/server.Dockerfile
+    container_name: catalogage-server
+    ports:
+      - "3579:3579"
+    environment:
+      APP_HOST: "0.0.0.0"
+      APP_DATABASE_URL: "postgresql+asyncpg://user:pass@db/catalogage"
+    restart: always
 
 volumes:
   pgdata:

--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -5,7 +5,7 @@ services:
     image: postgres:12
     container_name: catalogage-db
     ports:
-      - "5432:5432"
+      - "6432:5432"
     environment:
       - POSTGRES_DB=catalogage
       - POSTGRES_USER=user

--- a/docker/client.Dockerfile
+++ b/docker/client.Dockerfile
@@ -1,0 +1,13 @@
+FROM node:16 AS base
+WORKDIR /app/
+COPY ./client/package.json ./client/package-lock.json ./client/
+
+FROM base AS dependencies
+RUN cd client && npm ci
+
+FROM base AS release
+COPY --from=dependencies /app/client/node_modules ./client/node_modules
+COPY . .
+EXPOSE 3000
+WORKDIR /app/client
+CMD ["npm", "run", "dev", "--", "--host"]

--- a/docker/server.Dockerfile
+++ b/docker/server.Dockerfile
@@ -1,0 +1,14 @@
+FROM python:3.8 AS base
+WORKDIR /app/
+
+FROM base AS dependencies
+COPY ./requirements.txt ./
+RUN python3 -m venv venv && \
+    venv/bin/pip install -U pip wheel setuptools && \
+    venv/bin/pip install -r requirements.txt
+
+FROM base AS release
+COPY --from=dependencies /app/venv ./venv
+COPY . .
+EXPOSE 3579
+CMD ["venv/bin/python", "-m", "server.main"]

--- a/docs/fr/demarrage.md
+++ b/docs/fr/demarrage.md
@@ -2,6 +2,47 @@
 
 Cette page indique les quelques étapes qui vous permettront d'avoir un projet fonctionnel à partir duquel travailler.
 
+* 👉 Je suis une personne non-technique et je souhaite rapidement lancer le projet : jetez un oeil à [Raccourci : usage Docker](#raccourci--usage-docker)
+* 👉 Je suis une personne technique et je souhaite participer au développement détaillé : suivez le guide !
+
+**Table des matières**
+
+* [Usage Docker](#raccourci--usage-docker)
+* [Prérequis](#pr%C3%A9requis)
+* [Interagir avec le projet](#interagir-avec-le-projet)
+* [Configuration](#configuration)
+
+---
+
+## Usage Docker
+
+Pour faciliter un démarrage rapide du projet, une configuration `docker-compose` est disponible, comprenant le serveur, le client, et une base de données.
+
+Installez d'abord les outils nécessaires :
+
+- [Docker](https://docs.docker.com/get-docker/)
+- [Docker Compose](https://docs.docker.com/compose/install/)
+
+Puis lancez :
+
+```
+make up
+```
+
+Le client (application web) sera alors disponible sur http://localhost:3000.
+
+Le client se rechargera automatiquement après des modifications dans le dossier `client/src/`, ce qui vous permettra d'y faire des modifications et de voir rapidement le résultat.
+
+Pour toute autre modification, il faudra relancer Docker Compose : `make down` puis `make up`.
+
+Pour arrêter le Docker Compose :
+
+```
+make down
+```
+
+---
+
 ## Prérequis
 
 Selon que vous vouliez contribuer au serveur ou au client, vous allez avoir besoin de :
@@ -26,13 +67,22 @@ Si vous avez un [serveur PostgreSQL](https://www.postgresql.org/download/linux/)
 createdb catalogage
 ```
 
-Sinon, vous pouvez utiliser [Docker Compose](https://docs.docker.com/compose/install/) :
+Sinon, vous pouvez utiliser la configuration `docker-compose` (voir [Raccourci : usage Docker](#raccourci--usage-docker)) :
 
 ```bash
-docker-compose up
+docker-compose up -d -- db
 ```
 
-Assurez-vous de [configurer](#configuration) votre `APP_DATABASE_URL` en conséquence.
+[Configurez](#configuration) ensuite votre `APP_DATABASE_URL` :
+
+```
+cp .env.example .env
+```
+
+```bash
+# .env
+APP_DATABASE_URL="postgresql+asyncpg://user:pass@localhost:5432/${DB:-catalogage}"
+```
 
 ## Interagir avec le projet
 
@@ -97,35 +147,29 @@ make check
 
 ## Configuration
 
-Le serveur d'API est configurable à l'aide des variables d'environnement suivantes.
+Le projet est configurable à l'aide des variables d'environnement suivantes.
 
 | Variable | Description | Valeur par défaut |
 |---|---|---|
-| `APP_SERVER_MODE` | Un mode d'opération qui configure Uvicorn en conséquence : <br> - `local` : pour le développement local (_hot reload_ activé, etc) <br> - `live` : pour tout déploiement tel que défini via Ansible (voir [Opérations](./ops.md)) | `local` |
 | `APP_DATABASE_URL` | URL vers la base de données PostgreSQL | `postgresql+asyncpg://localhost:5432/catalogage` |
-| `APP_PORT` | Port du server d'API | `3579` |
-| `VITE_API_BROWSER_URL` | (1) URL utilisée par le navigateur lors de requêtes d'API | `http://localhost:3579` |
-| `VITE_API_SSR_URL` | (1) URL utilisée par le serveur frontend lors de requêtes d'API | `http://localhost:3579` |
 
-> Notes : 
->
-> * (1) Ces options n'existent que pour la configuration de production, où le service est déployé via Nginx. Elles ne doivent typiquement pas être modifiées lors du développement local.
+Définissez les valeurs spécifiques à votre situation dans un fichier `.env` placé à la racine du projet, que vous pouvez créer à partir du modèle `.env.example` :
 
-Définissez les valeurs spécifiques à votre situation dans un fichier `.env` placé à la racine du projet, que vous pouvez créer à partir du modèle `.env.example`.
+```
+cp .env .env.example
+```
 
 Les variables peuvent aussi être passées en arguments, elles sont alors utilisées en priorité par rapport au `.env`. Par exemple :
 
 ```bash
-APP_DATABASE_URL="..." make serve
+APP_DATABASE_URL="postgresql+asyncpg://..." make serve
 ```
 
-## Tests
+Des paramètres avancés (principalement dédiés au déploiement - voir [Opérations](./ops.md)) sont également disponibles :
 
-### Tests unitaires - Client
-
-Les tests unitaires côté client utilisent [`svelte-testing-library`](https://github.com/testing-library/svelte-testing-library).
-
-Autres ressources pour démarrer :
-
-- [Svelte Testing Library: Example](https://testing-library.com/docs/svelte-testing-library/example)
-- [Unit Testing Svelte Components](https://sveltesociety.dev/recipes/testing-and-debugging/unit-testing-svelte-component/)
+| Variable | Description | Valeur par défaut |
+|---|---|---|
+| `APP_SERVER_MODE` | Un mode d'opération qui configure Uvicorn en conséquence : <br> - `local` : pour le développement local (_hot reload_ activé, etc) <br> - `live` : pour tout déploiement tel que défini via Ansible | `local` |
+| `APP_PORT` | Port du server d'API | `3579` |
+| `VITE_API_BROWSER_URL` | URL utilisée par le navigateur lors de requêtes d'API. En mode `live`, indiquer le chemin vers l'API configuré sur Nginx : `/api`. | `http://localhost:3579` |
+| `VITE_API_SSR_URL` | URL utilisée par le serveur frontend lors de requêtes d'API | `http://localhost:3579` |

--- a/docs/fr/demarrage.md
+++ b/docs/fr/demarrage.md
@@ -2,12 +2,12 @@
 
 Cette page indique les quelques étapes qui vous permettront d'avoir un projet fonctionnel à partir duquel travailler.
 
-* 👉 Je suis une personne non-technique et je souhaite rapidement lancer le projet : jetez un oeil à [Raccourci : usage Docker](#raccourci--usage-docker)
+* 👉 Je suis une personne non-technique et je souhaite rapidement lancer le projet : jetez un oeil à [Usage Docker](#usage-docker)
 * 👉 Je suis une personne technique et je souhaite participer au développement détaillé : suivez le guide !
 
 **Table des matières**
 
-* [Usage Docker](#raccourci--usage-docker)
+* [Usage Docker](#usage-docker)
 * [Prérequis](#pr%C3%A9requis)
 * [Interagir avec le projet](#interagir-avec-le-projet)
 * [Configuration](#configuration)
@@ -67,7 +67,7 @@ Si vous avez un [serveur PostgreSQL](https://www.postgresql.org/download/linux/)
 createdb catalogage
 ```
 
-Sinon, vous pouvez utiliser la configuration `docker-compose` (voir [Raccourci : usage Docker](#raccourci--usage-docker)) :
+Sinon, vous pouvez utiliser la configuration `docker-compose` (voir [Usage Docker](#usage-docker)) :
 
 ```bash
 docker-compose up -d -- db

--- a/docs/fr/demarrage.md
+++ b/docs/fr/demarrage.md
@@ -26,19 +26,19 @@ Installez d'abord les outils nécessaires :
 Puis lancez :
 
 ```
-make up
+make compose-up
 ```
 
 Le client (application web) sera alors disponible sur http://localhost:3000.
 
 Le client se rechargera automatiquement après des modifications dans le dossier `client/src/`, ce qui vous permettra d'y faire des modifications et de voir rapidement le résultat.
 
-Pour toute autre modification, il faudra relancer Docker Compose : `make down` puis `make up`.
+Pour toute autre modification, il faudra relancer Docker Compose : `make compose-down` puis `make compose-up`.
 
 Pour arrêter le Docker Compose :
 
 ```
-make down
+make compose-down
 ```
 
 ---

--- a/docs/fr/outils.md
+++ b/docs/fr/outils.md
@@ -44,12 +44,23 @@ Ce projet est équipé du _type checking_ avec [`mypy`](https://mypy.readthedocs
 
 La vérification se fait avec une exécution standard de `$ mypy` lors de `make check`.
 
-## Playwright
+## Tests
+
+### Tests E2E
 
 Les tests _end-to-end_ sont lancés avec [`Playwright`](https://playwright.dev/).
 Il sont soit exécutés en mode _ci_ (continuous integration, par exemple dans
 notre cas avec github actions), et donc en _headless_, soit de manière
 interactive (en dev).
+
+### Tests unitaires - Client
+
+Les tests unitaires côté client utilisent [`svelte-testing-library`](https://github.com/testing-library/svelte-testing-library).
+
+Autres ressources pour démarrer :
+
+- [Svelte Testing Library: Example](https://testing-library.com/docs/svelte-testing-library/example)
+- [Unit Testing Svelte Components](https://sveltesociety.dev/recipes/testing-and-debugging/unit-testing-svelte-component/)
 
 ## DSFR
 

--- a/server/api/app.py
+++ b/server/api/app.py
@@ -3,7 +3,6 @@ from fastapi.middleware.cors import CORSMiddleware
 
 from server.config import Settings
 from server.config.di import resolve
-from server.infrastructure.database import Database
 
 from .routes import router
 
@@ -13,11 +12,9 @@ origins = [
 
 
 def create_app() -> FastAPI:
-    db = resolve(Database)
     settings = resolve(Settings)
 
     app = FastAPI(
-        on_startup=[db.create_all],
         docs_url=settings.docs_url,
     )
     app.add_middleware(

--- a/server/config/settings.py
+++ b/server/config/settings.py
@@ -11,6 +11,7 @@ class Settings(BaseSettings):
 
     server_mode: ServerMode = "local"
     database_url: str = "postgresql+asyncpg://localhost:5432/catalogage"
+    host: str = "localhost"
     port: int = 3579
     docs_url: str = "/docs"
     testing: bool = False

--- a/server/infrastructure/database.py
+++ b/server/infrastructure/database.py
@@ -13,9 +13,5 @@ class Database:
             self._engine, autocommit=False, autoflush=False, class_=AsyncSession
         )
 
-    async def create_all(self) -> None:
-        async with self._engine.begin() as conn:
-            await conn.run_sync(Base.metadata.create_all)
-
     def session(self) -> AsyncSession:
         return self._session_cls()

--- a/server/main.py
+++ b/server/main.py
@@ -14,7 +14,7 @@ if __name__ == "__main__":
     settings = resolve(Settings)
 
     kwargs: dict = {
-        "host": "127.0.0.1",
+        "host": settings.host,
         "port": settings.port,
     }
 

--- a/tools/wait-for-it.sh
+++ b/tools/wait-for-it.sh
@@ -1,0 +1,183 @@
+#!/usr/bin/env bash
+# Copied from: https://github.com/vishnubob/wait-for-it
+# Use this script to test if a given TCP host/port are available
+
+WAITFORIT_cmdname=${0##*/}
+
+echoerr() { if [[ $WAITFORIT_QUIET -ne 1 ]]; then echo "$@" 1>&2; fi }
+
+usage()
+{
+    cat << USAGE >&2
+Usage:
+    $WAITFORIT_cmdname host:port [-s] [-t timeout] [-- command args]
+    -h HOST | --host=HOST       Host or IP under test
+    -p PORT | --port=PORT       TCP port under test
+                                Alternatively, you specify the host and port as host:port
+    -s | --strict               Only execute subcommand if the test succeeds
+    -q | --quiet                Don't output any status messages
+    -t TIMEOUT | --timeout=TIMEOUT
+                                Timeout in seconds, zero for no timeout
+    -- COMMAND ARGS             Execute command with args after the test finishes
+USAGE
+    exit 1
+}
+
+wait_for()
+{
+    if [[ $WAITFORIT_TIMEOUT -gt 0 ]]; then
+        echoerr "$WAITFORIT_cmdname: waiting $WAITFORIT_TIMEOUT seconds for $WAITFORIT_HOST:$WAITFORIT_PORT"
+    else
+        echoerr "$WAITFORIT_cmdname: waiting for $WAITFORIT_HOST:$WAITFORIT_PORT without a timeout"
+    fi
+    WAITFORIT_start_ts=$(date +%s)
+    while :
+    do
+        if [[ $WAITFORIT_ISBUSY -eq 1 ]]; then
+            nc -z $WAITFORIT_HOST $WAITFORIT_PORT
+            WAITFORIT_result=$?
+        else
+            (echo -n > /dev/tcp/$WAITFORIT_HOST/$WAITFORIT_PORT) >/dev/null 2>&1
+            WAITFORIT_result=$?
+        fi
+        if [[ $WAITFORIT_result -eq 0 ]]; then
+            WAITFORIT_end_ts=$(date +%s)
+            echoerr "$WAITFORIT_cmdname: $WAITFORIT_HOST:$WAITFORIT_PORT is available after $((WAITFORIT_end_ts - WAITFORIT_start_ts)) seconds"
+            break
+        fi
+        sleep 1
+    done
+    return $WAITFORIT_result
+}
+
+wait_for_wrapper()
+{
+    # In order to support SIGINT during timeout: http://unix.stackexchange.com/a/57692
+    if [[ $WAITFORIT_QUIET -eq 1 ]]; then
+        timeout $WAITFORIT_BUSYTIMEFLAG $WAITFORIT_TIMEOUT $0 --quiet --child --host=$WAITFORIT_HOST --port=$WAITFORIT_PORT --timeout=$WAITFORIT_TIMEOUT &
+    else
+        timeout $WAITFORIT_BUSYTIMEFLAG $WAITFORIT_TIMEOUT $0 --child --host=$WAITFORIT_HOST --port=$WAITFORIT_PORT --timeout=$WAITFORIT_TIMEOUT &
+    fi
+    WAITFORIT_PID=$!
+    trap "kill -INT -$WAITFORIT_PID" INT
+    wait $WAITFORIT_PID
+    WAITFORIT_RESULT=$?
+    if [[ $WAITFORIT_RESULT -ne 0 ]]; then
+        echoerr "$WAITFORIT_cmdname: timeout occurred after waiting $WAITFORIT_TIMEOUT seconds for $WAITFORIT_HOST:$WAITFORIT_PORT"
+    fi
+    return $WAITFORIT_RESULT
+}
+
+# process arguments
+while [[ $# -gt 0 ]]
+do
+    case "$1" in
+        *:* )
+        WAITFORIT_hostport=(${1//:/ })
+        WAITFORIT_HOST=${WAITFORIT_hostport[0]}
+        WAITFORIT_PORT=${WAITFORIT_hostport[1]}
+        shift 1
+        ;;
+        --child)
+        WAITFORIT_CHILD=1
+        shift 1
+        ;;
+        -q | --quiet)
+        WAITFORIT_QUIET=1
+        shift 1
+        ;;
+        -s | --strict)
+        WAITFORIT_STRICT=1
+        shift 1
+        ;;
+        -h)
+        WAITFORIT_HOST="$2"
+        if [[ $WAITFORIT_HOST == "" ]]; then break; fi
+        shift 2
+        ;;
+        --host=*)
+        WAITFORIT_HOST="${1#*=}"
+        shift 1
+        ;;
+        -p)
+        WAITFORIT_PORT="$2"
+        if [[ $WAITFORIT_PORT == "" ]]; then break; fi
+        shift 2
+        ;;
+        --port=*)
+        WAITFORIT_PORT="${1#*=}"
+        shift 1
+        ;;
+        -t)
+        WAITFORIT_TIMEOUT="$2"
+        if [[ $WAITFORIT_TIMEOUT == "" ]]; then break; fi
+        shift 2
+        ;;
+        --timeout=*)
+        WAITFORIT_TIMEOUT="${1#*=}"
+        shift 1
+        ;;
+        --)
+        shift
+        WAITFORIT_CLI=("$@")
+        break
+        ;;
+        --help)
+        usage
+        ;;
+        *)
+        echoerr "Unknown argument: $1"
+        usage
+        ;;
+    esac
+done
+
+if [[ "$WAITFORIT_HOST" == "" || "$WAITFORIT_PORT" == "" ]]; then
+    echoerr "Error: you need to provide a host and port to test."
+    usage
+fi
+
+WAITFORIT_TIMEOUT=${WAITFORIT_TIMEOUT:-15}
+WAITFORIT_STRICT=${WAITFORIT_STRICT:-0}
+WAITFORIT_CHILD=${WAITFORIT_CHILD:-0}
+WAITFORIT_QUIET=${WAITFORIT_QUIET:-0}
+
+# Check to see if timeout is from busybox?
+WAITFORIT_TIMEOUT_PATH=$(type -p timeout)
+WAITFORIT_TIMEOUT_PATH=$(realpath $WAITFORIT_TIMEOUT_PATH 2>/dev/null || readlink -f $WAITFORIT_TIMEOUT_PATH)
+
+WAITFORIT_BUSYTIMEFLAG=""
+if [[ $WAITFORIT_TIMEOUT_PATH =~ "busybox" ]]; then
+    WAITFORIT_ISBUSY=1
+    # Check if busybox timeout uses -t flag
+    # (recent Alpine versions don't support -t anymore)
+    if timeout &>/dev/stdout | grep -q -e '-t '; then
+        WAITFORIT_BUSYTIMEFLAG="-t"
+    fi
+else
+    WAITFORIT_ISBUSY=0
+fi
+
+if [[ $WAITFORIT_CHILD -gt 0 ]]; then
+    wait_for
+    WAITFORIT_RESULT=$?
+    exit $WAITFORIT_RESULT
+else
+    if [[ $WAITFORIT_TIMEOUT -gt 0 ]]; then
+        wait_for_wrapper
+        WAITFORIT_RESULT=$?
+    else
+        wait_for
+        WAITFORIT_RESULT=$?
+    fi
+fi
+
+if [[ $WAITFORIT_CLI != "" ]]; then
+    if [[ $WAITFORIT_RESULT -ne 0 && $WAITFORIT_STRICT -eq 1 ]]; then
+        echoerr "$WAITFORIT_cmdname: strict mode, refusing to execute subprocess"
+        exit $WAITFORIT_RESULT
+    fi
+    exec "${WAITFORIT_CLI[@]}"
+else
+    exit $WAITFORIT_RESULT
+fi

